# [\#6881 PR](https://github.com/ethereum/solidity/pull/6881) `closed`: Fix assertion failure that happend when codegen the conversion procedure of converting from calldata array

#### <img src="https://avatars.githubusercontent.com/u/10496191?v=4" width="50">[sifmelcara](https://github.com/sifmelcara) opened issue at [2019-06-03 11:25](https://github.com/ethereum/solidity/pull/6881):

### Description

Fix #6835 

Runtime test work in progress

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages


#### <img src="https://avatars.githubusercontent.com/u/10496191?v=4" width="50">[sifmelcara](https://github.com/sifmelcara) commented at [2019-06-03 11:33](https://github.com/ethereum/solidity/pull/6881#issuecomment-498220404):

The bug is: When `convertType()` tries to recursively convert nested calldata array, it forgot to deal with the case that `typeOnStack` is calldata. (Probably because nested calldata array is not a thing before ABIEncoderV2)

#### <img src="https://avatars.githubusercontent.com/in/254?v=4" width="50">[codecov](https://github.com/apps/codecov) commented at [2019-06-03 12:08](https://github.com/ethereum/solidity/pull/6881#issuecomment-498229942):

# [Codecov](https://codecov.io/gh/ethereum/solidity/pull/6881?src=pr&el=h1) Report
> :exclamation: No coverage uploaded for pull request base (`develop@579bdad`). [Click here to learn what that means](https://docs.codecov.io/docs/error-reference#section-missing-base-commit).
> The diff coverage is `77.77%`.

[![Impacted file tree graph](https://codecov.io/gh/ethereum/solidity/pull/6881/graphs/tree.svg?width=650&token=87PGzVEwU0&height=150&src=pr)](https://codecov.io/gh/ethereum/solidity/pull/6881?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##             develop    #6881   +/-   ##
==========================================
  Coverage           ?   86.84%           
==========================================
  Files              ?      433           
  Lines              ?    41253           
  Branches           ?     4847           
==========================================
  Hits               ?    35827           
  Misses             ?     3825           
  Partials           ?     1601
```

| Flag | Coverage Δ | |
|---|---|---|
| #all | `86.84% <77.77%> (?)` | |
| #syntax | `25.18% <0%> (?)` | |

#### <img src="https://avatars.githubusercontent.com/u/10496191?v=4" width="50">[sifmelcara](https://github.com/sifmelcara) commented at [2019-06-19 09:46](https://github.com/ethereum/solidity/pull/6881#issuecomment-503489284):

I moved part of this PR's content to #6965 , as the clean up part and refactor part can be reviewed separately

#### <img src="https://avatars.githubusercontent.com/u/10496191?v=4" width="50">[sifmelcara](https://github.com/sifmelcara) commented at [2019-06-20 18:21](https://github.com/ethereum/solidity/pull/6881#issuecomment-504130988):

Hmm... Actually this PR only fixes `calldata -> memory` conversion. I just discovered that `calldata -> storage` conversion is also broken. I guess these two issue are independent though, looks like there is something wrong in `ArrayUtils::copyArrayToStorage()`. I will try to find the problem tomorrow or this weekend if I have time.

#### <img src="https://avatars.githubusercontent.com/u/10496191?v=4" width="50">[sifmelcara](https://github.com/sifmelcara) commented at [2019-06-20 18:26](https://github.com/ethereum/solidity/pull/6881#issuecomment-504132440):

Also it is strange that test suite throws exception if I add this function call to the test
```
// f(uint256[][])  : 0x20, 3, 0x60, 0xC0, 0x140, 2, 123, 124, 3, 223, 224, 225, 2, 323, 324 -> 32, 480, 0x20, 3, 0x60, 0xC0, 0x140, 2, 123, 124, 3, 223, 224, 225, 2, 323, 324
```
```
test/boostTest.cpp(125): error: in "semanticTests/abiEncoderV2/storage_array_encoding": Exception during extracted test: Throw location unknown (consider using BOOST_THROW_EXCEPTION)
Dynamic exception type: langutil::Error
std::exception::what: Literal encoding invalid.
[dev::tag_comment*] = Literal encoding invalid.
```
Maybe I missed something obvious?

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2019-06-24 12:07](https://github.com/ethereum/solidity/pull/6881#issuecomment-504981778):

The `0x140` is the problem.

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2019-06-24 12:07](https://github.com/ethereum/solidity/pull/6881#issuecomment-504981881):

Is this ready for review or not?

#### <img src="https://avatars.githubusercontent.com/u/10496191?v=4" width="50">[sifmelcara](https://github.com/sifmelcara) commented at [2019-06-24 12:19](https://github.com/ethereum/solidity/pull/6881#issuecomment-504985571):

Yes

On Mon, Jun 24, 2019, 8:08 PM chriseth <notifications@github.com> wrote:

> Is this ready for review or not?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ethereum/solidity/pull/6881?email_source=notifications&email_token=ACQCRP6C3HL5W6FCKOGGRTDP4C2K3A5CNFSM4HSHCWMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYMWS6I#issuecomment-504981881>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACQCRP5RSUSTRQMOWQUKAGDP4C2K3ANCNFSM4HSHCWMA>
> .
>

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2019-06-24 12:27](https://github.com/ethereum/solidity/pull/6881#issuecomment-504988259):

Looks good, but here, I would also like @ekpyron to take another look.

#### <img src="https://avatars.githubusercontent.com/u/10496191?v=4" width="50">[sifmelcara](https://github.com/sifmelcara) commented at [2019-07-05 08:20](https://github.com/ethereum/solidity/pull/6881#issuecomment-508675387):

@chriseth could you explain why `0x140` is the problem? It might not be possible to avoid using larger number like `0x140` if we want 3 dimensional array in semantic tests.

@ekpyron What cases only occur with three-dimensional array? (Maybe I overlook some side-effects happend when recursively converting sub-arrays) If you provide an example, I can trace the code to see if it functions correctly before I start writing SolidityEndToEndTests.

#### <img src="https://avatars.githubusercontent.com/u/1347491?v=4" width="50">[ekpyron](https://github.com/ekpyron) commented at [2019-07-10 15:03](https://github.com/ethereum/solidity/pull/6881#issuecomment-510100031):

@sifmelcara 
* In the test expectations hex literals are parsed as in solidity and require byte-padding, i.e. you need to use ``0x0140`` instead of ``0x140``.
* An example of a case that only occurs in three dimensional arrays is arrays with dynamically encoded, but statically sized base types. I'm not necessarily saying that these cases won't work, but in any case they should be tested.

#### <img src="https://avatars.githubusercontent.com/u/10496191?v=4" width="50">[sifmelcara](https://github.com/sifmelcara) commented at [2019-07-10 20:19](https://github.com/ethereum/solidity/pull/6881#issuecomment-510213183):

@ekpyron 

* Ok I opened #7082 to track the issue.
* Actually those cases indeed won't work in #6983 ... Fortunately you reviewed these PRs.

#### <img src="https://avatars.githubusercontent.com/u/2992036?u=4eca2cc0900689644626077e900530b38f65f1f5&v=4" width="50">[fubar](https://github.com/fubar) commented at [2019-08-17 21:21](https://github.com/ethereum/solidity/pull/6881#issuecomment-522271457):

@sifmelcara @ekpyron can you give an update on where this is at? We ran into the same issue with multi-dimensional calldata bytes arrays. Thanks!

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2019-08-19 09:18](https://github.com/ethereum/solidity/pull/6881#issuecomment-522485838):

@fubar we are currently not actively working on this because we have to re-check that decoding and length checks for all kinds of calldata variables works properly.

#### <img src="https://avatars.githubusercontent.com/u/2992036?u=4eca2cc0900689644626077e900530b38f65f1f5&v=4" width="50">[fubar](https://github.com/fubar) commented at [2019-08-19 18:24](https://github.com/ethereum/solidity/pull/6881#issuecomment-522697338):

@chriseth is that work tracked in a github issue or PR? I'd like to be able to follow the progress.

#### <img src="https://avatars.githubusercontent.com/u/9073706?v=4" width="50">[chriseth](https://github.com/chriseth) commented at [2019-08-28 14:54](https://github.com/ethereum/solidity/pull/6881#issuecomment-525782545):

@fubar the ones linked here.

#### <img src="https://avatars.githubusercontent.com/u/424752?u=2d50de05ec528b9b84f8b905a56e90669b0f8927&v=4" width="50">[Marenz](https://github.com/Marenz) commented at [2019-11-05 15:32](https://github.com/ethereum/solidity/pull/6881#issuecomment-549874927):

The linked issue is closed and this needs rebasing. Any other steps required to move this forward?

#### <img src="https://avatars.githubusercontent.com/u/1347491?v=4" width="50">[ekpyron](https://github.com/ekpyron) commented at [2019-11-05 16:15](https://github.com/ethereum/solidity/pull/6881#issuecomment-549893498):

We decided that we don't want to properly implement this for the old ABIEncoder, but instead want to push ABIEncoderV2 forward, so I'm closing this. I'll open a new issue for improving error messages in these cases.


-------------------------------------------------------------------------------



[Export of Github issue for [ethereum/solidity](https://github.com/ethereum/solidity). Generated on 2024.12.15 at 06:45:24.]
